

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>functions.gene_stats &mdash; Gene Expression Analysis 0.0.1 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="../../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
        <script type="text/javascript" src="../../_static/jquery.js"></script>
        <script type="text/javascript" src="../../_static/underscore.js"></script>
        <script type="text/javascript" src="../../_static/doctools.js"></script>
        <script type="text/javascript" src="../../_static/language_data.js"></script>
    
    <script type="text/javascript" src="../../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../index.html" class="icon icon-home"> Gene Expression Analysis
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../modules.html">Source</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../Bioinformatics Resources.html">Bioinformatics Resources</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../Gene Analysis with Python.html">Gene Analysis with Python</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../Data Analysis with Python.html">Data Analysis with Python</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../Pytest Resources.html">Pytest Resources</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../Python Resources.html">Python Resources</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">Gene Expression Analysis</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../index.html">Docs</a> &raquo;</li>
        
          <li><a href="../index.html">Module code</a> &raquo;</li>
        
      <li>functions.gene_stats</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for functions.gene_stats</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">from</span> <span class="nn">sklearn.decomposition</span> <span class="k">import</span> <span class="n">PCA</span>
<span class="kn">from</span> <span class="nn">sklearn.preprocessing</span> <span class="k">import</span> <span class="n">StandardScaler</span><span class="p">,</span> <span class="n">quantile_transform</span>
<span class="kn">from</span> <span class="nn">itertools</span> <span class="k">import</span> <span class="n">chain</span>
<span class="kn">from</span> <span class="nn">functools</span> <span class="k">import</span> <span class="n">reduce</span>


<div class="viewcode-block" id="pca_calculation"><a class="viewcode-back" href="../../functions.html#functions.gene_stats.pca_calculation">[docs]</a><span class="k">def</span> <span class="nf">pca_calculation</span><span class="p">(</span><span class="n">pca_dfs</span><span class="p">:</span> <span class="p">[</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">]</span> <span class="ow">or</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;PCA on a DataFrame or on a list of DataFrames.</span>

<span class="sd">    Perform PCA on several datasets to determine extent of batch effects or to compare global differences between</span>
<span class="sd">    datasets. A DataFrame containing samples to compare or a list of DataFrames (usually to compare between samples for</span>
<span class="sd">    batch effect) is passed. Data is standardized using the sklearn.preprocessing.StandardScaler class. (n/2)-1 components</span>
<span class="sd">    are calculated from standardized data. DataFrames passed should be formatted with experiments as columns and genes</span>
<span class="sd">    as rows.</span>

<span class="sd">    References: https://towardsdatascience.com/pca-using-python-scikit-learn-e653f8989e60</span>
<span class="sd">                https://jakevdp.github.io/PythonDataScienceHandbook/05.09-principal-component-analysis.html</span>
<span class="sd">                http://www.marcottelab.org/users/BCH391L_2015/NBT_primer_PCA.pdf</span>


<span class="sd">    Args:</span>
<span class="sd">        pca_dfs: A DataFrame, or list of DataFrames containing gene expression data for PCA comparison.</span>

<span class="sd">    Returns: A DataFrame containing the first (n/2)-1 components</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">PCs</span> <span class="o">=</span> <span class="nb">int</span><span class="p">((</span><span class="nb">len</span><span class="p">(</span><span class="n">pca_dfs</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">to_list</span><span class="p">())</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">pca_dfs</span><span class="p">,</span> <span class="nb">list</span><span class="p">):</span>
        <span class="n">cols</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="nb">enumerate</span><span class="p">(</span><span class="n">chain</span><span class="p">(</span><span class="o">*</span><span class="p">[</span><span class="n">i</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">values</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">pca_dfs</span><span class="p">]))))</span>
        <span class="n">vals</span> <span class="o">=</span> <span class="p">[</span><span class="n">cols</span><span class="o">.</span><span class="n">values</span><span class="p">()]</span>
        <span class="n">vals</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">chain</span><span class="o">.</span><span class="n">from_iterable</span><span class="p">(</span><span class="n">vals</span><span class="p">))</span>
        <span class="n">pooled</span> <span class="o">=</span> <span class="p">[</span><span class="n">pca_dfs</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">pca_dfs</span><span class="p">[</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">])</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">pca_dfs</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)]</span>
        <span class="p">[</span><span class="n">i</span><span class="o">.</span><span class="n">dropna</span><span class="p">()</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">pooled</span><span class="p">]</span>
        <span class="n">merged</span> <span class="o">=</span> <span class="n">reduce</span><span class="p">(</span><span class="k">lambda</span> <span class="n">left</span><span class="p">,</span> <span class="n">right</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">left</span><span class="p">,</span> <span class="n">right</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="s1">&#39;gene&#39;</span><span class="p">),</span> <span class="n">pca_dfs</span><span class="p">)</span>
        <span class="n">pca</span> <span class="o">=</span> <span class="n">PCA</span><span class="p">(</span><span class="n">n_components</span><span class="o">=</span><span class="nb">int</span><span class="p">((</span><span class="nb">len</span><span class="p">(</span><span class="n">pca_dfs</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">))</span>

    <span class="k">else</span><span class="p">:</span>
        <span class="n">vals</span> <span class="o">=</span> <span class="n">pca_dfs</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">to_list</span><span class="p">()</span>
        <span class="n">i</span> <span class="o">=</span> <span class="nb">iter</span><span class="p">(</span><span class="n">vals</span><span class="p">)</span>
        <span class="n">cols</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">vals</span><span class="p">)),</span> <span class="n">i</span><span class="p">))</span>
        <span class="n">merged</span> <span class="o">=</span> <span class="n">pca_dfs</span>
        <span class="n">merged</span><span class="o">.</span><span class="n">dropna</span><span class="p">()</span>
        <span class="n">pca</span> <span class="o">=</span> <span class="n">PCA</span><span class="p">(</span><span class="n">n_components</span><span class="o">=</span><span class="n">PCs</span><span class="p">)</span>
    <span class="n">transposed</span> <span class="o">=</span> <span class="n">merged</span><span class="o">.</span><span class="n">transpose</span><span class="p">()</span>
    <span class="c1"># StandardScalar calculates z-score z = (x-mean)/std-dev</span>
    <span class="n">scaled</span> <span class="o">=</span> <span class="n">StandardScaler</span><span class="p">()</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">transposed</span><span class="p">)</span>
    <span class="n">pca_transformed</span> <span class="o">=</span> <span class="n">pca</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">scaled</span><span class="p">)</span>
    <span class="n">final_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">pca_transformed</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">vals</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Principal Component </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">PCs</span><span class="p">)])</span>
    <span class="k">return</span> <span class="n">final_df</span><span class="p">,</span> <span class="n">cols</span></div>


<div class="viewcode-block" id="quantile_norm"><a class="viewcode-back" href="../../functions.html#functions.gene_stats.quantile_norm">[docs]</a><span class="k">def</span> <span class="nf">quantile_norm</span><span class="p">(</span><span class="n">dfs</span><span class="p">:</span> <span class="p">[</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">]</span> <span class="ow">or</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Perform quantile normalization on expression data.</span>

<span class="sd">    Quantile normalize a list of DataFrames using sklearn.Preprocessing.quantile_transform with default n_quantiles=1000.</span>
<span class="sd">    The DataFrame(s) passed should contain samples from the same batch. Each DataFrame should be formatted with columns</span>
<span class="sd">    as samples and rows as genes.</span>

<span class="sd">    TODO: as of now the function assumes the gene order and number are the same across samples. This should change</span>

<span class="sd">    Args:</span>
<span class="sd">        dfs: a DataFrame or list of DataFrames from the same batch</span>

<span class="sd">    Returns: a dataframe containing</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">transformed</span> <span class="o">=</span> <span class="nb">list</span><span class="p">()</span>
    <span class="n">vals</span> <span class="o">=</span> <span class="p">[</span><span class="nb">list</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">values</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">dfs</span><span class="p">]</span>
    <span class="n">index</span> <span class="o">=</span> <span class="n">dfs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">dfs</span><span class="p">,</span> <span class="nb">list</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">dfs</span><span class="p">)):</span>
            <span class="n">transformed</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">quantile_transform</span><span class="p">(</span><span class="n">dfs</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">),</span> <span class="n">columns</span><span class="o">=</span><span class="n">vals</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">index</span><span class="o">=</span><span class="n">index</span><span class="p">))</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">transformed</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">quantile_transform</span><span class="p">(</span><span class="n">dfs</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">),</span> <span class="n">columns</span><span class="o">=</span><span class="n">vals</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">index</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">transformed</span></div>


<div class="viewcode-block" id="batch_effect_adjust"><a class="viewcode-back" href="../../functions.html#functions.gene_stats.batch_effect_adjust">[docs]</a><span class="k">def</span> <span class="nf">batch_effect_adjust</span><span class="p">(</span><span class="n">df</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Adjust for batch effects using R function ComBat from sva package</span>
<span class="sd">    Ref: https://www.youtube.com/watch?v=z3vqrkRGSLI</span>

<span class="sd">    Args:</span>
<span class="sd">        df:</span>

<span class="sd">    Returns:</span>

<span class="sd">    &quot;&quot;&quot;</span></div>


<div class="viewcode-block" id="calc_adj_pvalue"><a class="viewcode-back" href="../../functions.html#functions.gene_stats.calc_adj_pvalue">[docs]</a><span class="k">def</span> <span class="nf">calc_adj_pvalue</span><span class="p">(</span><span class="n">dfs</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Calculate adjusted p-value</span>

<span class="sd">    Args:</span>
<span class="sd">        dfs:</span>

<span class="sd">    Returns:</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">pass</span></div>


<span class="c1"># def kmeans_cluster(df):</span>
<span class="c1">#     &quot;&quot;&quot;</span>
<span class="c1">#     A DataFrame containing several samples</span>
<span class="c1">#     Args:</span>
<span class="c1">#         df:</span>
<span class="c1">#</span>
<span class="c1">#     Returns:</span>
<span class="c1">#     &quot;&quot;&quot;</span>
</pre></div>

           </div>
           
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2019, Colton Garelli &amp; Jillian Richmond

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>