

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>functions.modify_dataframes &mdash; Gene Expression Analysis 0.0.1 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="../../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
        <script type="text/javascript" src="../../_static/jquery.js"></script>
        <script type="text/javascript" src="../../_static/underscore.js"></script>
        <script type="text/javascript" src="../../_static/doctools.js"></script>
        <script type="text/javascript" src="../../_static/language_data.js"></script>
    
    <script type="text/javascript" src="../../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../index.html" class="icon icon-home"> Gene Expression Analysis
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../modules.html">Source</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../Bioinformatics Resources.html">Bioinformatics Resources</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../Gene Analysis with Python.html">Gene Analysis with Python</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../Data Analysis with Python.html">Data Analysis with Python</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../Pytest Resources.html">Pytest Resources</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../Python Resources.html">Python Resources</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../Python Resources.html#statements">Statements</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../Python Resources.html#data-types">Data Types</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../Python Resources.html#uniprotio">UniprotIO:</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../Python Resources.html#seqrecord-class">SeqRecord Class</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../Python Resources.html#biosql">BioSQL:</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../Python Resources.html#other-imported-modules">Other Imported Modules</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../Python Resources.html#misc">Misc</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">Gene Expression Analysis</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../index.html">Docs</a> &raquo;</li>
        
          <li><a href="../index.html">Module code</a> &raquo;</li>
        
      <li>functions.modify_dataframes</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for functions.modify_dataframes</h1><div class="highlight"><pre>
<span></span><span class="c1"># pd.concat([df3, df.get(&#39;time&#39;), df.get(&#39;40&#39;)], axis=1) to concat columns</span>
<span class="c1"># values = new_df.loc[~new_df[&#39;40&#39;].isin(range(-1,1))] to exclude things b/t -1,1</span>
<span class="c1"># pvals.append(float(decimal.Decimal(random.randrange(1, 1000))/10000))</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>


<div class="viewcode-block" id="value_cutoff"><a class="viewcode-back" href="../../functions.html#functions.modify_dataframes.value_cutoff">[docs]</a><span class="k">def</span> <span class="nf">value_cutoff</span><span class="p">(</span><span class="n">df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">col_name</span><span class="p">,</span> <span class="n">upper</span><span class="p">,</span> <span class="n">lower</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Removes rows containing values above an upper bound or **within** a range (upper, lower)...(NON-INCLUSIVE)</span>

<span class="sd">    A df containing data is passed in. The function removes the entire row when a value contained in the specified column</span>
<span class="sd">    (col_name) falls within the range. For logFC DEG analysis, upper and lower bounds should be provided. Only a lower</span>
<span class="sd">    bound is required for p-values (will remove everything &gt;lower bound).</span>


<span class="sd">    This function is intended to reduce memory load, specifically for efficient, non-laggy Plotly plots.</span>

<span class="sd">    Args:</span>
<span class="sd">        upper: An upper bound to eliminate genes from the passed df</span>
<span class="sd">        lower: A lower bound to eliminate genes from the passed df</span>
<span class="sd">        df: A df containing desired data to modify</span>
<span class="sd">        col_name: The column containing the data the function will</span>

<span class="sd">    Returns: A Dataframe containing only the filtered data</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">lower</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="c1"># everything less than the upperbound</span>
        <span class="n">new_df</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="n">col_name</span><span class="p">]</span> <span class="o">&lt;</span> <span class="n">upper</span><span class="p">]</span>
        <span class="k">return</span> <span class="n">new_df</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">new_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="n">col_name</span><span class="p">]</span> <span class="o">&lt;</span> <span class="n">lower</span><span class="p">])</span>
        <span class="n">new_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">new_df</span><span class="p">,</span> <span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="n">col_name</span><span class="p">]</span> <span class="o">&gt;</span> <span class="n">upper</span><span class="p">]])</span>
        <span class="k">return</span> <span class="n">new_df</span></div>


<div class="viewcode-block" id="find_common_genes"><a class="viewcode-back" href="../../functions.html#functions.modify_dataframes.find_common_genes">[docs]</a><span class="k">def</span> <span class="nf">find_common_genes</span><span class="p">(</span><span class="n">frames</span><span class="p">:</span> <span class="p">[</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">]):</span>
    <span class="sd">&quot;&quot;&quot;Takes a list of DataFrames and returns a DataFrame containing genes common between all DataFrames in the list.</span>

<span class="sd">    Filter out uncommon genes from a set of DataFrame. Useful for comparing gene targets from different</span>
<span class="sd">    tissues/cell types/diseases where constitutive expression may vary.</span>
<span class="sd">    Args:</span>
<span class="sd">        frames: a list of Pandas DataFrames containing transcription data</span>

<span class="sd">    Returns:</span>
<span class="sd">        A Pandas DataFrame only with genes that overlap between all DataFrames in frames</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Drop values that don&#39;t have shared genes (symbol column)</span>
    <span class="n">common</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">i1</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">frames</span><span class="p">)</span><span class="o">-</span><span class="mi">1</span><span class="p">):</span>
        <span class="n">counter1</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">counter2</span> <span class="o">=</span> <span class="mi">1</span>
        <span class="k">if</span> <span class="n">i1</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">common</span> <span class="o">=</span> <span class="n">frames</span><span class="p">[</span><span class="n">i1</span><span class="p">]</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">frames</span><span class="p">[</span><span class="n">i1</span><span class="o">+</span><span class="mi">1</span><span class="p">],</span> <span class="n">lsuffix</span><span class="o">=</span><span class="nb">str</span><span class="p">(</span><span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">counter1</span><span class="p">))),</span>
                                     <span class="n">rsuffix</span><span class="o">=</span><span class="nb">str</span><span class="p">(</span><span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">counter2</span><span class="p">))),</span> <span class="n">how</span><span class="o">=</span><span class="s1">&#39;inner&#39;</span><span class="p">)</span>
            <span class="n">counter1</span> <span class="o">+=</span> <span class="mi">2</span>
            <span class="n">counter2</span> <span class="o">+=</span> <span class="mi">2</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">temp</span> <span class="o">=</span> <span class="n">common</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">deep</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="n">temp</span> <span class="o">=</span> <span class="n">temp</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">frames</span><span class="p">[</span><span class="n">i1</span><span class="o">+</span><span class="mi">1</span><span class="p">],</span> <span class="n">how</span><span class="o">=</span><span class="s1">&#39;inner&#39;</span><span class="p">,</span> <span class="n">lsuffix</span><span class="o">=</span><span class="nb">str</span><span class="p">(</span><span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">counter1</span><span class="p">))),</span>
                             <span class="n">rsuffix</span><span class="o">=</span><span class="nb">str</span><span class="p">(</span><span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">counter2</span><span class="p">))))</span>
            <span class="n">counter1</span> <span class="o">+=</span> <span class="mi">2</span>
            <span class="n">counter2</span> <span class="o">+=</span> <span class="mi">2</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">temp</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
                <span class="n">common</span> <span class="o">=</span> <span class="n">temp</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">deep</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="c1"># If nothing in common between two of the frames</span>
                <span class="c1"># Return empty frame</span>
                <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>
    <span class="k">return</span> <span class="n">common</span></div>


<div class="viewcode-block" id="drop_cols"><a class="viewcode-back" href="../../functions.html#functions.modify_dataframes.drop_cols">[docs]</a><span class="k">def</span> <span class="nf">drop_cols</span><span class="p">(</span><span class="n">df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">cols_to_drop</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Drop columns from a list of column names.</span>


<span class="sd">    TODO: Duplicate of file_in.strip_data</span>
<span class="sd">    Args:</span>
<span class="sd">        df:</span>
<span class="sd">        cols_to_drop:</span>

<span class="sd">    Returns:</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">df</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="n">cols_to_drop</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></div>


<div class="viewcode-block" id="drop_genes"><a class="viewcode-back" href="../../functions.html#functions.modify_dataframes.drop_genes">[docs]</a><span class="k">def</span> <span class="nf">drop_genes</span><span class="p">(</span><span class="n">df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">genes</span><span class="p">:</span> <span class="nb">list</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Drops undesired genes from a DataFrame</span>

<span class="sd">    Args:</span>
<span class="sd">        df:</span>
<span class="sd">        genes:</span>

<span class="sd">    Returns:</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">return_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span>
    <span class="n">count</span> <span class="o">=</span> <span class="mi">00</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">():</span>
        <span class="k">if</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">genes</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">count</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">temp</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[[</span><span class="n">i</span><span class="p">]]</span>
                <span class="n">return_df</span> <span class="o">=</span> <span class="n">temp</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">temp</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[[</span><span class="n">i</span><span class="p">]]</span>
                <span class="n">return_df</span> <span class="o">=</span> <span class="n">return_df</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">temp</span><span class="p">)</span>
            <span class="n">count</span> <span class="o">+=</span> <span class="mi">1</span>
    <span class="k">return</span> <span class="n">return_df</span></div>


<div class="viewcode-block" id="remove_duplicate_indicies"><a class="viewcode-back" href="../../functions.html#functions.modify_dataframes.remove_duplicate_indicies">[docs]</a><span class="k">def</span> <span class="nf">remove_duplicate_indicies</span><span class="p">(</span><span class="n">df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Keeps first instance of a gene/index and removes the others</span>

<span class="sd">    Args:</span>
<span class="sd">        df:</span>

<span class="sd">    Returns:</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="o">~</span><span class="n">df</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">duplicated</span><span class="p">(</span><span class="n">keep</span><span class="o">=</span><span class="s1">&#39;first&#39;</span><span class="p">)]</span></div>

</pre></div>

           </div>
           
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2019, Colton Garelli &amp; Jillian Richmond

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>